Partial verification secondary

tanishqjasoria opened the pull request:
### Purpose of the PR:
Implement and demonstrate the partial verification Secondary.
Addressing the issue #48 

### Summary of Changes:
- Update the structure of the partial metadata distributed by the Primary
- Update demo file to comply with the partial verification Secondary also
- Add function to verify the partial metadata, compling to the latest Uptane standard
- Update sample metadata to include an update for the new partial verification secondary
- Update test_secondary to perform tests for the partial verification Secondary also
    - Verifying metadata and retrieving the target fileinfo
    - Validating the image for the partial verifying secondary

### Further Requirements:
The Readme files needed to be updated

tanishqjasoria committed:
Update to be compatible with partial verification

tanishqjasoria committed:
Update metadata distributed to partial verification secondaries

Now metadata distributed to the partial verification secondaries
would be in the form of a archive with the structure almost
similar to that of the distributed archive for full verification.
Here only director repo would be present and it would only have
the targets metadata.

This is done so that the secondary could obtain, save and expand
the metadata recieved from primary in a similar fashion as that
of full verification. It would further help in simplifying the
partial verification process

tanishqjasoria committed:
Update to reciver the metadata for partial verification

Now primary is sending archive for partial verification
metadata as well.

tanishqjasoria committed:
Add function for Partial Verification of Secondary

Used various tuf low level functions to verify the targets metadata
supplied by the director repository. It complies with the latest
Uptane standard for design.

tanishqjasoria committed:
Update to store last validated targets until new found

Earlier even if no target_info was provided the
validated_targets_for_this_ecu would be emptied.

Now if new target info is present, then the list
validated_targets_for_this_ecu would be updated

tanishqjasoria committed:
Restructure the test data in the test_secondary

Earlier the test data was stores in seperate lists like
TEMP_CLIENT_DIRS, vins, ecu_serials and secondary_instances

Now all the data corresponding to a particular secondary is
stored in a dictionary and all the dictionaries are further
stored in a list TEST_INSTANCES

tanishqjasoria committed:
Update the test verifying the parital_metadata generaton for secondary

Update the function get_partial_metadata_fname to
get_partial_metadata_archive_fname

tanishqjasoria committed:
Update samples to accomodate the changes in partial metadata

Earlier the partial metadata supplied by the primary was
just director_targets file but now the primary provides a
archive with the structure similar to that of full metadata

tanishqjasoria committed:
Update metadata files to include update for the partial verifying ECU

Modified metadata files to include a possible update
for the newly included partial verifying secondary ECU.

tanishqjasoria committed:
Add two more TEST_INSTANCES for the partial verifying secondary ECUs

tanishqjasoria committed:
Add pv__expected_updated_fileinfo to make partial verifying ECU accept the update

This included fileinfo corresponds to the info of the
update that the partial verifying ECU would donwnload
as per the new metadata.

tanishqjasoria committed:
Modify to include the tests for partial verifying secondary ECU

The tests for partial verification ECU are carried out in the
similar fashion as that of the full verification ECU.

tanishqjasoria committed:
FIX error in test_20_update_time due to wrong function call

tanishqjasoria committed:
Add new sample metadata files for tests to work with der format

Due to some discrepency in json and der metadata
the tests failed while working with der formats

tanishqjasoria committed:
Modify test to validate partial verification Secondary firmware also

lukpueh commented on tests/test_secondary.py:
```diff
@@ -57,15 +57,54 @@
+# client directory,  whether or not the instance is partial-verifying, the
+# vehicle's ID, the Secondary's ID, and a reference to the instance.
+# Also note the nonce we'll use when validating sample time attestation data.
+# Changing the nonce or  would require producing new signed test data
```
or?

tanishqjasoria replied:
I have edited the comment. It was incomplete earlier!

lukpueh commented on tests/test_secondary.py:
```diff
@@ -57,15 +57,54 @@
+        'partial_verifying': True,
+        'vin': '000',
+        'ecu_serial': 'pv_00000',
+        'instance': None}]
```
Minor: These test instances are often addressed by their index. For the sake of readability, would it make sense to make `TEST_INSTANCES` a dictionary (ordered, if necessary), so each instance can be addressed by a descriptive name (if such a name is available)?

tanishqjasoria replied:
I can keep their name as that of their client_directory?

lukpueh commented on tests/test_secondary.py:
```diff
@@ -77,7 +116,7 @@
         'length': 37}}
 
-expected_updated_fileinfo = {
+fv_expected_updated_fileinfo = {
```
Minor: Could you either expand `fv` (and `pv`, here and elsewhere in this PR) or add a comment that explains the abbreviation? Even though I'm reviewing a "partial verification" PR, it took me a few seconds to realize what the initials stand for.

lukpueh commented on tests/test_secondary.py:
```diff
@@ -685,41 +758,60 @@ def test_40_process_metadata(self):
+      if instance_data['partial_verifying']:
+        repo_list = ['director']
+        # Both root and targets as there is root file needed to establish a root
+        # of trust while shipping the ECU
```
Could you please explain this comment?

tanishqjasoria replied:
I have changed the comment!
I wanted to state that, in partial verification secondary, root file would also be present as a root of trust when the firmware of the ECU is packaged by the vendor!

lukpueh commented on uptane/clients/secondary.py:
```diff
@@ -564,8 +566,117 @@ def fully_validate_metadata(self):
+
+
+    <Arguments>
+      director_targets_metadata_fname
```
This function does not have an argument.

tanishqjasoria replied:
I restructured the function and forgot to change the doc string afterwards

lukpueh commented on uptane/clients/secondary.py:
```diff
@@ -564,8 +566,117 @@ def fully_validate_metadata(self):
+
+    <Exceptions>
+      uptane.Error
+        if director_targets_metadata_fname does not specify a file that exists
```
There is `director_targets_metadata_fname`

lukpueh commented on uptane/clients/secondary.py:
```diff
@@ -564,8 +566,117 @@ def fully_validate_metadata(self):
+    #  to a minimal value.  It will be updated in the next cycle.
+
+    # Is the metadata is not expired?
+    director_obj._ensure_not_expired(director_obj.metadata['current']['targets'], 'targets')
```
line len

lukpueh commented on uptane/clients/secondary.py:
```diff
@@ -564,8 +566,117 @@ def fully_validate_metadata(self):
+      validated_targets_from_director = []
+
+    # Get the targets specified by the role itself.
+    for filepath, fileinfo in six.iteritems(director_obj.metadata['current']['targets']['targets']):
```
line len

lukpueh commented on uptane/clients/secondary.py:
```diff
@@ -564,8 +566,117 @@ def fully_validate_metadata(self):
+
+    # _update_metadata carries out the verification of metadata
+    # and then updates the metadata of the repository
+    director_obj._update_metadata('targets', upperbound_filelength)
```
This seems crucial. It might be worth mentioning that `_update_metadata` downloads, verifies, and 'installs' the director metadata and, if it successful, that the current and previous metadata stores are updated.

lukpueh commented on uptane/clients/secondary.py:
```diff
@@ -564,8 +566,117 @@ def fully_validate_metadata(self):
+  def partial_validate_metadata(self):
+    """
+    <Purpose>
+      Given the filename of a file containing the Director's Targets role
```
Where does the `filename` come from?

lukpueh commented on uptane/clients/secondary.py:
```diff
@@ -564,8 +566,117 @@ def fully_validate_metadata(self):
+    # Do we have metadata for 'targets'?
+    if 'targets' not in director_obj.metadata['current']:
+      log.debug('No metadata for \'targtes\'. Unable to determine targets.')
+      validated_targets_from_director = []
```
Why re-assign an empty list to `validated_targets_from_director`?

lukpueh replied:
`director_obj.metadata['current']['targets']` in the for loop below will blow up if `'targets' not in director_obj.metadata['current']` ever evaluates to true here (which btw. it does not in your unit tests).

lukpueh replied:
Maybe you want to `return` here in the if block?

lukpueh commented on uptane/clients/secondary.py:
```diff
@@ -564,8 +566,117 @@ def fully_validate_metadata(self):
+      new_target['filepath'] = filepath
+      new_target['fileinfo'] = fileinfo
+
+      validated_targets_from_director.append(new_target)
```
`validated_targets_from_director` gets discarded at the end of this function without *really* being needed. 

Suggestion: You could scrap the second loop, move up the `if 'custom' not in ...` and fill `validated_targets_for_this_ecu` right here. And to assign it then to `self.validated_targets_for_this_ecu`, depending on whether there are `validated_targets_from_director` (as you do below), a boolean (e.g. `is_validated_targets_from_director` would suffice.

 This might be a matter of taste but IMO it makes the code simpler.

lukpueh commented on uptane/clients/secondary.py:
```diff
@@ -564,8 +566,117 @@ def fully_validate_metadata(self):
+      validated_targets_for_this_ecu.append(target)
+
+    if validated_targets_from_director:
+      self.validated_targets_for_this_ecu = validated_targets_for_this_ecu
```
Could you leave a comment that explains why you look at `validated_targets_from_director` to decide whether you update `self.validated_targets_for_this_ecu`? 

For comparison, in `fully_validate_metadata` above you look at `validated_targets_for_this_ecu`.

Btw. the Secondary class-docstring about `self.validated_targets_for_this_ecu` seems to need an update.

lukpueh commented on uptane/clients/secondary.py:
```diff
@@ -602,18 +713,56 @@ def get_validated_target_info(self, target_filepath):
+    as long as it checks out as trustworthy.
+
+    Full:
+      The given filename, metadata_fname, should point to an archive of all
```
s/metadata_fname/metadata_archive_fname ?

lukpueh commented on uptane/clients/secondary.py:
```diff
@@ -602,18 +713,56 @@ def get_validated_target_info(self, target_filepath):
+      underlying TUF code.
+
+    Partial:
+      The given filename, metadata_fname, should point to a single metadata
```
s/metadata_fname/metadata_archive_fname ?

lukpueh commented on uptane/clients/secondary.py:
```diff
@@ -602,18 +713,56 @@ def get_validated_target_info(self, target_filepath):
+
+    Partial:
+      The given filename, metadata_fname, should point to a single metadata
+      role file, the Director's Targets role. The signature on the Targets role
```
Minor nit: Does it really point to a single metadata file, or rather to an archive with a [repository structure](https://github.com/uptane/uptane/blob/0dd0403cab7f6d7f804fc149a324b52f1e53c0b5/uptane/clients/primary.py#L1294) with a single file in it?

tanishqjasoria replied:
It points to the archive with a repository structure with a single file in it. I would update this!

lukpueh commented on uptane/clients/secondary.py:
```diff
@@ -602,18 +713,56 @@ def get_validated_target_info(self, target_filepath):
+      role file, the Director's Targets role. The signature on the Targets role
+      file is validated against the Director's public key
+      (self.director_public_key). If the signature is valid, the new Targets
+      role file is trusted, else it is discarded and we TUF raises a
```
-we

lukpueh commented on uptane/clients/primary.py:
```diff
@@ -1267,23 +1265,43 @@ def save_distributable_metadata_files(self):
+      # tuf.conf.METADATA_FORMAT (presumably .json or .der, depending on
+      # that setting).
+      if not os.path.isfile(role_abs_fname) or not role_abs_fname.endswith(
+          '.' + tuf.conf.METADATA_FORMAT):
```
Given the two statements before, I think there is no need to check if `not role_abs_fname.endswith('.' + tuf.conf.METADATA_FORMAT)`.

lukpueh commented on uptane/clients/primary.py:
```diff
@@ -1267,23 +1265,43 @@ def save_distributable_metadata_files(self):
+        # Consider special error type.
+        raise uptane.Error('Unexpected file type in a metadata '
+            'directory: ' + repr(role_abs_fname) + ' Expecting only ' +
+            tuf.conf.METADATA_FORMAT + 'files.')
```
Missing space before `files.'`

lukpueh commented on demo/demo_secondary.py:
```diff
@@ -63,6 +63,7 @@
 _ecu_serial = 'TCUdemocar'
 _primary_host = demo.PRIMARY_SERVER_HOST
 _primary_port = demo.PRIMARY_SERVER_DEFAULT_PORT
+_partial_verifying = False
```
Have you tried running the [demo](https://github.com/uptane/uptane/#1-starting-the-demo) with `_partial_verifying = True`?

tanishqjasoria replied:
Yes, I did!

lukpueh reviewed (changes requested):
Thanks for this PR, @tanishqjasoria! This looks mostly good, judging from #48 and the corresponding parts in the [Uptane Standard](https://uptane.github.io/uptane-standard/uptane-standard.html#rfc.section.5.4.4.1). I left requests for clarification, where I wasn't sure, and raised a few minor concerns here and there. Note that I did not review any metadata.

Please let me know when you think this is ready for another round of review.

tanishqjasoria committed:
Fix issues with formatting and comments

tanishqjasoria committed:
Edit docstring of partial_validate_metadata

tanishqjasoria committed:
Restructure partial_validate_metadata

tanishqjasoria committed:
Reorganize TEST_INSTANCES as dictionaries

Earlier all the 6 test instances were stored
as list. Now all the test instances are stored
as a dictionaries indexed by their client directory
names
